iotdb-core/ainode/ainode/core/inference/inference_request_pool.py

```diff
@@ -65,65 +67,88 @@ def __init__(
         self._waiting_queue = request_queue  # Requests that are waiting to be processed
         self._running_queue = mp.Queue()  # Requests that are currently being processed
         self._finished_queue = result_queue  # Requests that are finished
+        self._scheduler = BasicScheduler(
+            self._waiting_queue, self._running_queue, self._finished_queue
+        )
         self._stop_event = mp.Event()
 
         # Fix inference seed
         random.seed(self.FIX_SEED)
         torch.manual_seed(self.FIX_SEED)
         np.random.seed(self.FIX_SEED)
 
-    def memory_is_available(self, request):
-        # need test with several rounds of dummy data
-        pass
+    def _warm_up_and_estimate_memory(self):
+        # TODO: Test per token memory usage
+        torch.cuda.empty_cache()
+        gc.collect()
+        dummy_input = torch.zeros(
+            (1, self.config.input_token_len), dtype=torch.float32
+        ).to(self.device)
+
+        # force cuda synchronization to avoid any asynchronous memory allocation issues
+        torch.cuda.reset_peak_memory_stats(self.device)
+        torch.cuda.synchronize(self.device)
+        memory_before_warmup = torch.cuda.memory_allocated(self.device)
+        logger.info(
+            f"Before warm-up, peak memory usage: {memory_before_warmup:.2f} bytes"
+        )
 
-    def _activate_requests(self):
-        if self._waiting_queue.empty():
-            return
-        request: InferenceRequest = self._waiting_queue.get()
-        # TODO: Check memory size before activating requests
-        request.inputs = request.inference_pipeline.preprocess_inputs(request.inputs)
-        request.mark_running()
-        logger.debug(
-            f"[Inference][Device-{self.device}][Pool-{self.pool_id}][ID-{request.req_id}] Request is activated with inputs shape {request.inputs.shape}"
+        # warm-up
+        with torch.no_grad():
+            self.model.generate(dummy_input, max_new_tokens=1)
+        torch.cuda.synchronize(self.device)
+        peak_memory_1_token = torch.cuda.max_memory_allocated(self.device)
+        logger.info(
+            f"Baseline memory usage for 1 token: {peak_memory_1_token:.2f} bytes"
+        )
+        logger.info(
+            f"Differentiation : {peak_memory_1_token-memory_before_warmup:.2f} bytes"
         )
-        self._running_queue.put(request)
+
+    def _activate_requests(self):
+        requests = self._scheduler.schedule_activate()
+        for request in requests:
+            request.inputs = request.inference_pipeline.preprocess_inputs(
+                request.inputs
+            )
```

Comment:
U should deliver the scheduled requests to running_queue here rather than within the scheduler. Otherwise, some requests would be `step` without preprocess their inputs.